### PR TITLE
Fix multiple chances of a usage reporting dialog NPE (fixes #1154)

### LINKED ISSUE #1154: Multiple chances of a usage reporting dialog NPE
AppV 0.10.11 (this bug is also expected to be present in ealier releases)

Error Log from GPlay Console
1) java.lang.NullPointerException: 
  at com.nutomic.syncthingandroid.service.RestApi.getOptions (RestApi.java:371)
  at com.nutomic.syncthingandroid.activities.MainActivity.lambda$showUsageReportingDialog$10$MainActivity (MainActivity.java:477)
  at com.nutomic.syncthingandroid.activities.MainActivity$$Lambda$9.onClick (Unknown Source)

2) java.lang.NullPointerException: 
  at com.nutomic.syncthingandroid.activities.MainActivity.lambda$showUsageReportingDialog$10$MainActivity (MainActivity.java:477)
  at com.nutomic.syncthingandroid.activities.MainActivity$$Lambda$9.onClick (Unknown Source)

Mentioned in issue list #1146

### PR DESCRIPTION
**Pull request by ghost**
Purpose
Fix async restApi.getSystemInfo requests causing the usage reporting dialog to throw NPE if syncthing shuts down until the user chooses an option. To avoid making three times a restApi.getOptions() and restApi.getSystemInfo -> { callback } in a row from MainActivity, I cached the mUrVersionMax needed to get the usage dialog decision within RestApi before it signals "Full class initialization and State.ACTIVE". To make the code shorter and more straight forward, I moved the code for "isUsageReportingDecided" and "setUsageReporting" to RestApi. This also ensures all needed variable stuff is present and accessible when we need it.

See issue #1154 for more details on the error report.

Testing
I verified this working correctly on device lg-h815 running Android 7.1.2 by setting the "show ur dialog time to 90 seconds instead of 3 days" temporarily at commit https://github.com/syncthing/syncthing-android/pull/1155/commits/c5ed92218995dfa8b4c1b4f3ba4001e9fc47025e .

### CODE REVIEW

File: app/src/main/java/com/nutomic/syncthingandroid/activities/MainActivity.java [hunk 1]
```diff
@@ -475,28 +477,36 @@ private void setOptimalDrawerWidth(View drawerContainer) {
...
-                    });
-                    break;
-                case DialogInterface.BUTTON_NEGATIVE:
-                    options.urAccepted = Options.USAGE_REPORTING_DENIED;
-                    break;
-                case DialogInterface.BUTTON_NEUTRAL:
-                    Uri uri = Uri.parse("https://data.syncthing.net");
-                    startActivity(new Intent(Intent.ACTION_VIEW, uri));
-                    break;
+            if (restApi == null) {
```
Comment (AudriusButkevicius) [inline app/src/main/java/com/nutomic/syncthingandroid/activities/MainActivity.java]: No point for this check as this does not get called if restApi is null in the outer stack.

File: app/src/main/java/com/nutomic/syncthingandroid/activities/MainActivity.java [hunk 2]
```diff
@@ -475,28 +477,36 @@ private void setOptimalDrawerWidth(View drawerContainer) {
...
+                    case DialogInterface.BUTTON_NEGATIVE:
+                        restApi.setUsageReporting(false);
+                        restApi.restart();
+                        break;
+                    case DialogInterface.BUTTON_NEUTRAL:
+                        Uri uri = Uri.parse("https://data.syncthing.net");
+                        startActivity(new Intent(Intent.ACTION_VIEW, uri));
+                        break;
+                }
+            } catch (Exception e) {
```
Comment (AudriusButkevicius) [inline app/src/main/java/com/nutomic/syncthingandroid/activities/MainActivity.java]: again, not a fan of the catch all situation, but I guess it's better than crashing.

File: app/src/main/java/com/nutomic/syncthingandroid/activities/MainActivity.java (hunk 1 above)
Comment (Catfriend1) [reply app/src/main/java/com/nutomic/syncthingandroid/activities/MainActivity.java]: Careful, this is callback code, so restApi may be null when the user "sits there some minutes and run conditions shut down syncthingservice + restApi in the meantime".

File: app/src/main/java/com/nutomic/syncthingandroid/activities/MainActivity.java [hunk 3]
```diff
@@ -475,28 +477,36 @@ private void setOptimalDrawerWidth(View drawerContainer) {
...
+                Log.e(TAG, "showUsageReportingDialog:OnClickListener", e);
             }
-            getApi().editSettings(getApi().getGui(), options);
-            getApi().restart();
         };
 
-        getApi().getUsageReport(report -> {
+        if (restApi == null) {
```
Comment (Catfriend1) [inline app/src/main/java/com/nutomic/syncthingandroid/activities/MainActivity.java]: We can omit this part as it's checked by the outer stack. I'll do so.

File: app/src/main/java/com/nutomic/syncthingandroid/activities/MainActivity.java (hunk 2 above)
Comment (Catfriend1) [reply app/src/main/java/com/nutomic/syncthingandroid/activities/MainActivity.java]: Yes, I agree with you on that.

File: app/src/main/java/com/nutomic/syncthingandroid/activities/MainActivity.java (hunk 1 above)
Comment (AudriusButkevicius) [reply app/src/main/java/com/nutomic/syncthingandroid/activities/MainActivity.java]: Well no, as rest API is passed in as an argument, and before being passed as an argument, there is an if that checks that it's not null.

Comment (Catfriend1) [reply app/src/main/java/com/nutomic/syncthingandroid/activities/MainActivity.java]: Ok, but as this is a delayed callback I suggest getting rid of https://github.com/syncthing/syncthing-android/pull/1155/commits/408e6dafc3f7f6c50ce7f993b5d0601d0cf50c53 instead.

File: app/src/main/java/com/nutomic/syncthingandroid/service/RestApi.java [hunk 4]
```diff
@@ -536,4 +542,23 @@ public void getUsageReport(final OnResultListener1<String> listener) {
...
+    }
+
+    public void setUsageReporting(Boolean acceptUsageReporting) {
+        Options options = getOptions();
+        if (options == null) {
+            Log.e(TAG, "setUsageReporting called while options == null");
+            return;
+        }
+        options.urAccepted = acceptUsageReporting ? mUrVersionMax : Options.USAGE_REPORTING_DENIED;
+        mConfig.options = options;
```
Comment (AudriusButkevicius) [inline app/src/main/java/com/nutomic/syncthingandroid/service/RestApi.java]: This is weird, why is this setting some mConfig.options as a side effect?

Comment (AudriusButkevicius) [reply app/src/main/java/com/nutomic/syncthingandroid/service/RestApi.java]: I can't see a save or anything anywhere.

Comment (Catfriend1) [reply app/src/main/java/com/nutomic/syncthingandroid/service/RestApi.java]: see MainActivity#showUsageReportingDialog for restApi.restart();
Maybe we should rename the restApi.restart() function to restApi.saveAndRestart() in a follow-up PR?

Comment (Catfriend1) [reply app/src/main/java/com/nutomic/syncthingandroid/service/RestApi.java]: For improvement on this matter, please see PR #1158

Comment (AudriusButkevicius) [reply app/src/main/java/com/nutomic/syncthingandroid/service/RestApi.java]: Still, I don't see a save or anything called.

File: app/src/main/java/com/nutomic/syncthingandroid/activities/MainActivity.java (hunk 1 above)
Comment (AudriusButkevicius) [reply app/src/main/java/com/nutomic/syncthingandroid/activities/MainActivity.java]: Sorry but how is this async?

image

Comment (Catfriend1) [reply app/src/main/java/com/nutomic/syncthingandroid/activities/MainActivity.java]: see line 481 in your screenshot. The onClickListener will fire some time later when the user clicked on the answer button in the dialog. So line 482 and so on are not guaranteed to have a valid restApi not being null available.

File: app/src/main/java/com/nutomic/syncthingandroid/service/RestApi.java (hunk 4 above)
Comment (Catfriend1) [reply app/src/main/java/com/nutomic/syncthingandroid/service/RestApi.java]: The save is here:
image
lines 490 and 494

File: app/src/main/java/com/nutomic/syncthingandroid/activities/MainActivity.java (hunk 1 above)
Comment (AudriusButkevicius) [reply app/src/main/java/com/nutomic/syncthingandroid/activities/MainActivity.java]: Well, you are guaranteed to have a non-null RestAPI, as that was passed to the function, and that's the rest api used in scope of the lambda.

Comment (Catfriend1) [reply app/src/main/java/com/nutomic/syncthingandroid/activities/MainActivity.java]: Ok, picked up your feedback in commit https://github.com/syncthing/syncthing-android/pull/1155/commits/c5ed92218995dfa8b4c1b4f3ba4001e9fc47025e

**Comment by Catfriend1** [comment]
@AudriusButkevicius Is there anything that still needs to be done I've overlooked? May I have your approval?

**Comment by Catfriend1** [comment]
Sorry I didn't want to bump, maybe you currently haven't the time to review this? So I'll wait some more ... :)

**Comment by AudriusButkevicius** [comment]
I'll get to it, I've barely been home.

File: app/src/main/java/com/nutomic/syncthingandroid/service/RestApi.java (hunk 4 above)
Comment (AudriusButkevicius) [reply app/src/main/java/com/nutomic/syncthingandroid/service/RestApi.java]: Yeah, let's rename the function in a separate PR.

Comment (Catfriend1) [reply app/src/main/java/com/nutomic/syncthingandroid/service/RestApi.java]: We already did this.

File: app/src/main/java/com/nutomic/syncthingandroid/activities/MainActivity.java [hunk 5]
```diff
@@ -475,28 +477,29 @@ private void setOptimalDrawerWidth(View drawerContainer) {
...
+                        break;
+                    case DialogInterface.BUTTON_NEUTRAL:
+                        Uri uri = Uri.parse("https://data.syncthing.net");
+                        startActivity(new Intent(Intent.ACTION_VIEW, uri));
+                        break;
+                }
+            } catch (Exception e) {
+                Log.e(TAG, "showUsageReportingDialog:OnClickListener", e);
             }
-            getApi().editSettings(getApi().getGui(), options);
```
[the remark this answers is not in the document]
Comment (Catfriend1) [reply app/src/main/java/com/nutomic/syncthingandroid/activities/MainActivity.java]: Line 486+490 use it. The marked line is being replaced by them as we don't need to save if the user does not decide anything.
I've already updated this PR to incorporate the new saveConfigAndRestart function.

[the remark this answers is not in the document]
Comment (Catfriend1) [reply app/src/main/java/com/nutomic/syncthingandroid/activities/MainActivity.java]: I'll check if editSettings is still used tomorrow. (Todo)

**Review by AudriusButkevicius — APPROVED** [review]

**Pull request merged**